### ISSUE TITLE
compute a KL divergence for a Gaussian Mixture prior and a normal distribution posterior

### ISSUE THREAD

**Issue by neuronphysics** [pytorch/botorch#1720]
# 🚀 Feature Request

Hi,

I am trying to compute a KL divergence for my variational autoencoder architecture between a Gaussian Mixture prior and a normal distribution posterior. Since this KL divergence is analytically intractable and only it could be evaluated by some sort of approximation, I am wondering whether it is also possible to compute it via Monte Carlo Sampling using `botorch`? I will appreciate if   you could suggest an implementation with your library? Here is the architecture of my encoder and prior networks 
```
import torch
import torch.nn as nn
from torch.distributions.normal import Normal
from torch.distributions import MixtureSameFamily, Categorical
from torch.distributions.independent import Independent
class VGMM(nn.Module): 
     def __init__(self,
                  u_dim,
                  h_dim,
                  z_dim,
                  n_mixtures,
                  batch_norm=False,
                  )
        super(VGMM, self).__init__()
        self.n_mixtures =n_mixtures
        self.batch_norm=  batch_norm
        encoder_layers=[nn.Linear(self.x_dim , self.h_dim)]
        if self.batch_norm:
            encoder_layers.append(torch.nn.BatchNorm1d(self.h_dim))
        encoder_layers=encoder_layers+[
            nn.ReLU(),
            nn.Linear(self.h_dim, self.h_dim),
        ]
        if self.batch_norm:
            encoder_layers= encoder_layers+[nn.BatchNorm1d(self.h_dim)]

        encoder_layers  = encoder_layers+[nn.ReLU()]

        self.enc        = torch.nn.Sequential(*encoder_layers)

        self.enc_mean   = nn.Linear(self.h_dim, self.z_dim)

        self.enc_logvar = nn.Linear(self.h_dim, self.z_dim)
        self.dist = MixtureSameFamily
        self.comp = Normal
        self.mix = Categorical

        layers_prior = [nn.Linear(self.u_dim, self.h_dim)]
        if self.batch_norm:
            layers_prior.append(torch.nn.BatchNorm1d(self.h_dim))
        layers_prior = layers_prior + [
            nn.ReLU(),
        ]

        self.prior = torch.nn.Sequential(*layers_prior)

        self.prior_mean = nn.ModuleList(
            [nn.Linear(self.h_dim, self.z_dim) for _ in range(n_mixtures)]
        )

        self.prior_logvar = nn.ModuleList(
            [nn.Linear(self.h_dim, self.z_dim) for _ in range(n_mixtures)]
        )
        self.prior_weights = nn.Linear(input_dim, n_mixtures) 

     def forward(self, u):
        encoder_input = self.enc(u)
        enc_mean  = self.enc_mean(encoder_input)
        enc_logvar= self.enc_logvar(encoder_input)
        enc_logvar = nn.Softplus()(enc_logvar)
        prior_mean= torch.cat([ self.prior_mean[n](u).unsqueeze(1) for n in range(self.n_mixtures)],dim=1,)
        prior_logvar = torch.cat([self.prior_logvar[n](u).unsqueeze(1)for n in range(self.n_mixtures)],dim=1,)
        prior_w = self.prior_weights(prior_input)
        prior_sigma = prior_logvar.exp().sqrt()
        prior_dist = self.dist(self.mix(logits=prior_w), Independent(self.comp(prior_mean, prior_sigma), 1))
        z_t      = self.reparametrization(enc_mean, enc_logvar)
        return prior_dist, z_t
     def reparametrization(self, mu, log_var):
        var = torch.exp(log_var* 0.5)
        eps = torch.FloatTensor(var.size()).normal_(mean=0, std=1).to(self.device)
        eps = torch.autograd.Variable(eps)
        return eps.mul(var).add_(mu).add_(1e-7)
```
Is there any possibility to use `botorch` library and expliciltly Monte Carlo sampling in order to compute this KL divergence? Thanks in advance.

**Comment by esantorella** [comment]
Hi, thanks for reaching out!

If I understand correctly, you’re asking about computing a KL divergence between two continuous distributions via Monte Carlo sampling. I would expect that it’s possible to use MCMC to do this, although I’m not sure how well it would converge.

However, BoTorch doesn’t have any functionality for estimating a KL divergence in that way, and since BoTorch is intended as a Bayesian Optimization library, we’re not planning on adding that functionality.

You can implement MCMC methods on top of BoTorch [Posterior](https://botorch.org/docs/posteriors) objects since you can sample from them, with gradients. However, you can also do this with the underlying torch distributions that you’re already using in the VAE, so I don’t think BoTorch would add a lot of value there.
